pkg/provisioner/ironic/ironic.go

```diff
@@ -1098,6 +1159,30 @@ func (p *ironicProvisioner) Adopt(force bool) (result provisioner.Result, err er
 	return
 }
 
+func (p *ironicProvisioner) ironicHasSameImage(ironicNode *nodes.Node) (sameImage bool) {
+	// To make it easier to test if ironic is configured with
+	// the same image we are trying to provision to the host.
+	if p.host.Spec.Image != nil && p.host.Spec.Image.DiskFormat != nil && *p.host.Spec.Image.DiskFormat == "live-iso" {
+		sameImage = (ironicNode.InstanceInfo["boot_iso"] == p.host.Spec.Image.URL)
```

Comment:
This is not cleared from Ironic when switching back from live-iso to a regular boot, so it may not be operative. I think we should either clear it at the same time as setting `deploy_interface` back to direct, or also verify the `deploy_interface` here, or both.

Reply:
Ah thanks, good catch - I updated `getImageUpdateOptsForNode` to account for this, and added tests for both image->iso and iso->image in `updateopts_test.go`

I'm not sure adding an explicit check for deploy_interface here gains us much though, since in that case we expect `*p.host.Spec.Image.DiskFormat` to change from `live-iso` to something else, then we'll hit the `else` and the checksum/image_source values in `ironicNode.InstanceInfo` will always be empty, so sameImage will always evaluate to false AFAICT?

That said, it does raise the question of whether we should check `*p.host.Spec.Image.DiskFormat` since in any other change of disk-format we might want to set sameImage to false, even if the URL is the same?  I guess that's something to potentially address in a different PR though.

Reply:
Agree that doing both is redundant. (Wouldn't hurt, in a belt-and-braces sense, but this is fine too.)

You make a good point though that if provisioning fails because you specified the wrong image format, and you then fix it, it currently won't retry, at least on paper. I'm not actually even 100% sure this code is still needed, since after a failure is first recorded we change the state to Deprovisioning. Definitely a question for elsewhere though.